larpandoracontent/LArCheating/CheatingBeamParticleIdTool.cc

```diff
@@ -0,0 +1,82 @@
+/**
+ *  @file   larpandoracontent/LArCheating/CheatingBeamParticleIdTool.cc
+ *
+ *  @brief  Implementation of the cheating beam particle id tool class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArCheating/CheatingBeamParticleIdTool.h"
+
+#include "larpandoracontent/LArHelpers/LArMCParticleHelper.h"
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+CheatingBeamParticleIdTool::CheatingBeamParticleIdTool() : 
+    m_minWeightFraction(0.5f)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void CheatingBeamParticleIdTool::SelectOutputPfos(const pandora::Algorithm *const /*pAlgorithm*/, const SliceHypotheses &nuSliceHypotheses, const SliceHypotheses &crSliceHypotheses, PfoList &selectedPfos)
+{
+    if (nuSliceHypotheses.size() != crSliceHypotheses.size())
+        throw StatusCodeException(STATUS_CODE_INVALID_PARAMETER);
+
+    IntVector beamSliceIndices;
+
+    for (unsigned int sliceIndex = 0, nSlices = nuSliceHypotheses.size(); sliceIndex < nSlices; ++sliceIndex)
+    {
+        float beamParticleWeight(0.f), totalWeight(0.f);
+        const PfoList &neutrinoPfoList(nuSliceHypotheses.at(sliceIndex));
+
+        for (const Pfo *const pNeutrinoPfo : neutrinoPfoList)
+        {
+            if (!LArPfoHelper::IsNeutrino(pNeutrinoPfo))
+                throw StatusCodeException(STATUS_CODE_INVALID_PARAMETER);
+
+            PfoList downstreamPfos;
+            LArPfoHelper::GetAllDownstreamPfos(pNeutrinoPfo, downstreamPfos);
+
+            float thisBeamParticleWeight(0.f), thisTotalWeight(0.f);
+            CheatingParticleIdBaseTool::GetTargetParticleWeight(&downstreamPfos, false, thisBeamParticleWeight, thisTotalWeight, LArMCParticleHelper::IsBeamParticle);
+
+            beamParticleWeight += thisBeamParticleWeight;
+            totalWeight += thisTotalWeight;
+        }
+
+        if (beamParticleWeight/totalWeight > m_minWeightFraction)
+        {
+            beamSliceIndices.push_back(sliceIndex);
```

Comment:
Shame that couldn't share this method with the neutrino case, but difference (single neutrino selected, vs. multiple beam particles) is clear.